core/src/main/scala/org/apache/spark/sql/TiStrategy.scala

```diff
@@ -185,18 +188,33 @@ class TiStrategy(context: SQLContext) extends Strategy with Logging {
     PredicateUtils.extractColumnRefFromExpression(expression).asScala.toSeq
 
   private def filterToDAGRequest(
+    projectList: Seq[NamedExpression],
     filters: Seq[Expression],
     source: TiDBRelation,
     dagRequest: TiDAGRequest = new TiDAGRequest(pushDownType(), timeZoneOffset())
   ): TiDAGRequest = {
-    val tiFilters: Seq[TiExpression] = filters.collect { case BasicExpression(expr) => expr }
+    val tiProjects: Seq[TiExpression] = projectList.collect { case BasicExpression(expr) => expr }
+    val tiFilters: Seq[TiExpression] = filters.collect { case BasicExpression(expr)      => expr }
+
+    val tiColumns
+      : Seq[TiColumnRef] = (tiFilters ++ tiProjects).flatMap { referencedTiColumns }.distinct
```

Comment:
Projections and filters should be extracted into `referencedTiColumns` so that we know which columns are needed in whole SQL.